src/DAG/dag_parser.c

```diff
@@ -2,10 +2,163 @@
+ demangle the keys in order to persist them.*/
+int _MangleTensorsNames(RedisModuleCtx *ctx, RedisAI_RunInfo *rinfo) {
+
+    int res = REDISMODULE_ERR;
+    AI_dict *mangled_tensors = AI_dictCreate(&AI_dictTypeHeapRStrings, NULL);
+
+    {
+        AI_dictIterator *iter = AI_dictGetSafeIterator(rinfo->dagTensorsContext);
+        AI_dictEntry *entry = AI_dictNext(iter);
+        while (entry) {
+            RedisModuleString *key = (RedisModuleString *)AI_dictGetKey(entry);
+            size_t key_len;
+            const char *key_str = RedisModule_StringPtrLen(key, &key_len);
+            RedisModuleString *demangled_key = RedisModule_CreateString(NULL, key_str, key_len - 4);
+            int *instance = RedisModule_Alloc(sizeof(int));
+            *instance = 1;
+            AI_dictAdd(mangled_tensors, (void *)demangled_key, (void *)instance);
+            RedisModule_FreeString(NULL, demangled_key);
+            entry = AI_dictNext(iter);
+        }
+        AI_dictReleaseIterator(iter);
+    }
+
+    for (long long i = 0; i < array_len(rinfo->dagOps); i++) {
+        RAI_DagOp *currentOp = rinfo->dagOps[i];
+
+        RedisModuleString **mangled_inkeys =
+            array_new(RedisModuleString *, array_len(currentOp->inkeys));
+        for (long long j = 0; j < array_len(currentOp->inkeys); j++) {
+            RedisModuleString *key = currentOp->inkeys[j];
+            AI_dictEntry *entry = AI_dictFind(mangled_tensors, key);
+            if (!entry) {
+                array_free(mangled_inkeys);
+                RedisModule_ReplyWithError(ctx, "ERR INPUT key cannot be found in DAG");
+                goto cleanup;
+            }
+            int *instance = AI_dictGetVal(entry);
+            char buf[16];
+            sprintf(buf, "%04d", *instance);
+            RedisModuleString *mangled_key = RedisModule_CreateStringFromString(NULL, key);
+            RedisModule_StringAppendBuffer(NULL, mangled_key, buf, strlen(buf));
+            mangled_inkeys = array_append(mangled_inkeys, mangled_key);
+        }
+
+        RedisModuleString **mangled_outkeys =
+            array_new(RedisModuleString *, array_len(currentOp->outkeys));
+        for (long long j = 0; j < array_len(currentOp->outkeys); j++) {
+            RedisModuleString *key = currentOp->outkeys[j];
+            AI_dictEntry *entry = AI_dictFind(mangled_tensors, key);
+            int *instance = NULL;
+            if (entry) {
+                instance = AI_dictGetVal(entry);
+                *instance += 1;
+            } else {
+                instance = RedisModule_Alloc(sizeof(int));
+                *instance = 1;
+                AI_dictAdd(mangled_tensors, (void *)key, (void *)instance);
+            }
+            char buf[16];
+            sprintf(buf, "%04d", *instance);
+            RedisModuleString *mangled_key = RedisModule_CreateStringFromString(NULL, key);
+            RedisModule_StringAppendBuffer(NULL, mangled_key, buf, strlen(buf));
```

Comment:
can you skip the creation? just append on the current key? this will prevent unnecessary allocations and freeing of the arrays